retrieval/discovery/kubernetes/discovery.go

```diff
@@ -735,3 +788,233 @@ func nodeHostIP(node *Node) (net.IP, error) {
+	}
+
+	var pods PodList
+	if err := json.NewDecoder(res.Body).Decode(&pods); err != nil {
+		body, _ := ioutil.ReadAll(res.Body)
+		return nil, "", fmt.Errorf("unable to list Kubernetes pods; unexpected response body: %s", string(body))
+	}
+
+	podMap := map[string]map[string]*Pod{}
+	for idx, pod := range pods.Items {
+		if _, ok := podMap[pod.ObjectMeta.Namespace]; !ok {
+			podMap[pod.ObjectMeta.Namespace] = map[string]*Pod{}
+		}
+		log.Debugf("Got pod %s in namespace %s", pod.ObjectMeta.Name, pod.ObjectMeta.Namespace)
+		podMap[pod.ObjectMeta.Namespace][pod.ObjectMeta.Name] = &pods.Items[idx]
+	}
+
+	return podMap, pods.ResourceVersion, nil
+}
+
+func (kd *Discovery) watchPods(events chan interface{}, done <-chan struct{}, retryInterval time.Duration) {
+	until(func() {
+		pods, resourceVersion, err := kd.getPods()
+		if err != nil {
+			log.Errorf("Cannot initialize pods collection: %s", err)
+			return
+		}
+		kd.podsMu.Lock()
+		kd.pods = pods
+		kd.podsMu.Unlock()
+
+		req, err := http.NewRequest("GET", podsURL, nil)
+		if err != nil {
+			log.Errorf("Cannot create pods request: %s", err)
+			return
+		}
+
+		values := req.URL.Query()
+		values.Add("watch", "true")
+		values.Add("resourceVersion", resourceVersion)
+		req.URL.RawQuery = values.Encode()
+		res, err := kd.queryAPIServerReq(req)
+		if err != nil {
+			log.Errorf("Failed to watch pods: %s", err)
+			return
+		}
+		defer res.Body.Close()
+		if res.StatusCode != http.StatusOK {
+			log.Errorf("Failed to watch pods: %d", res.StatusCode)
+			return
+		}
+
+		d := json.NewDecoder(res.Body)
+
+		for {
+			var event podEvent
+			if err := d.Decode(&event); err != nil {
+				log.Errorf("Watch pods unexpectedly closed: %s", err)
+				return
+			}
+
+			select {
+			case events <- &event:
+			case <-done:
+			}
+		}
+	}, retryInterval, done)
+}
+
+func podSource(pod *Pod) string {
+	return sourcePodPrefix + ":" + pod.ObjectMeta.Namespace + ":" + pod.ObjectMeta.Name
+}
+
+type ByContainerPort []ContainerPort
+
+func (a ByContainerPort) Len() int           { return len(a) }
+func (a ByContainerPort) Less(i, j int) bool { return a[i].ContainerPort < a[j].ContainerPort }
+func (a ByContainerPort) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+
+func updatePodTargets(pod *Pod, allContainers bool) []model.LabelSet {
+	var targets []model.LabelSet = make([]model.LabelSet, 0, len(pod.PodSpec.Containers))
+	if pod.PodStatus.PodIP == "" {
+		log.Debugf("skipping pod %s -- PodStatus.PodIP is empty", pod.ObjectMeta.Name)
+		return targets
+	}
+
+	if pod.PodStatus.Phase != "Running" {
+		log.Debugf("skipping pod %s -- status is not `Running`", pod.ObjectMeta.Name)
+		return targets
+	}
+
+	ready := "unknown"
+	for _, cond := range pod.PodStatus.Conditions {
+		if strings.ToLower(cond.Type) == "ready" {
+			ready = strings.ToLower(cond.Status)
+		}
+	}
+
+	for _, container := range pod.PodSpec.Containers {
```

Comment:
Is the ordering here deterministic, or can kubernetes change the order the user provides?